src/Mvc/Mvc.Analyzers/src/Startup/StartupAnalyzer.Diagnostics.cs

```diff
@@ -16,5 +16,15 @@ public partial class StartupAnalzyer : DiagnosticAnalyzer
             DiagnosticSeverity.Warning, 
             isEnabledByDefault: true,
             helpLinkUri: "https://aka.ms/YJggeFn");
+
+        internal readonly static DiagnosticDescriptor BuildServiceProviderShouldNotCalledInConfigureServicesMethod = new DiagnosticDescriptor(
+            "MVC1007",
+            "Cannot use BuildServiceProvider in ConfigureServices.",
+            "BuildServiceProvider called by runtime,don't call this method in your application code directly",
```

Comment:
Suggestion: `Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consider alternatives such as dependency injecting services as parameters to `Configure`.

The docs article should cover:
- DI of parameters to configure
- Using DI with options https://andrewlock.net/access-services-inside-options-and-startup-using-configureoptions/
- @davidfowl  anything else?

Reply:
Does this warn if you return the IServiceProvider?

Reply:
No,this warn if called `BuildServiceProvider` in 'ConfigureServices'.

It seems add validators such as `ConfigureServiceMethodSignatureValidator` and `ConfigureMethodSignatureValidator` to check startup methods signatures

Reply:
This only affects 3.0 applications right?

Reply:
These analyzers will only show up for 3.0+ because they ship in the SDK.